test/specs/jsCodegen/evaluateXPathWithJsCodegen.ts

```diff
@@ -0,0 +1,51 @@
+import {
+	compileXPathToJavaScript,
+	executeJavaScriptCompiledXPath,
+	IDomFacade,
+	IReturnTypes,
+	ReturnType,
+} from '../../../src/index';
+
+function generateKey(query: string, returnType: ReturnType) {
+	return `${query} ${returnType}`;
+}
+
+function normalizeEndOfLines(xpathString: string): string {
+	// Replace all character sequences of 0xD followed by 0xA and all 0xD not followed by 0xA with 0xA.
+	return xpathString.replace(/(\x0D+\x0A)|(\x0D+(?!\x0A))/g, String.fromCharCode(0xa));
+}
+
+const cache = {};
+
+const evaluateXPathWithJsCodegen = <
+	TNode extends Node,
+	TReturnType extends keyof IReturnTypes<TNode>
+>(
+	query: string,
+	contextItem?: any | null,
+	domFacade?: IDomFacade | null,
+	returnType?: ReturnType
+): IReturnTypes<TNode>[TReturnType] => {
+	returnType = returnType || (ReturnType.ANY as any);
+	query = normalizeEndOfLines(query);
+
+	const cachedQuery = cache[generateKey(query, returnType)];
+
+	if (!cachedQuery) {
+		const compiledXPathResult = compileXPathToJavaScript(query, returnType);
+		if (compiledXPathResult.isAstAccepted === true) {
```

Comment:
Let's see next Friday _why_ this is needed. There might be something in the typing of this object that we can improve.